### ISSUE TITLE
Generics

### ISSUE THREAD

**Issue by Ryanng-zr** [nus-cs2030/2021-s2#221]
Hi guys, I was wondering if Optional is a type, just like Integer, Double, String, etc. 
For example, using the Pair class as an example:
![image](https://user-images.githubusercontent.com/77624747/114858372-f7956200-9e1b-11eb-8b3d-142e1d89f8d3.png)

Can I do something like this: 
new Pair(Optional.ofNullable(...) , Optional.of(...))
Does Optional fit in S and T just like how Integer, Double and String does? And also, what types are S and T limited to?
Thanks a lot!

**Comment by rendao1010** [comment]
From my understanding, yes, it should work (although you should specify the types when initializing the new Pair, e.g. `new Pair<Optional<Integer>,Optional<Integer>>of(Optional.of(1), Optional.of(2))`).

S and T shouldn't be limited to any types - they can take on any class (unless you bound them in your class definition)!

**Comment by kuro1511** [comment]
i think S and T can be any Object? Then since optional is an object, it should work

**Comment by wglee99** [comment]
Yes, it should work. However, you should careful when you handle the retrieving of individual values from the pair (in which case would be Optionals) in the Aggregate class.

**Comment by rljw** [comment]
Yup it would work. From what I understand, Optionals are just wrapper classes that wraps around your objects S and T (whatever you define them to be), and they are of Object type, so it should work fine. The only difference is that the handling of the optional objects may be different from that of objects (have to use optional methods in java api to deal with them, handling of null values slightly different as well)

**Comment by gableejh** [comment]
it should work, optional works just like a wrapper which wraps the object and value within in

**Comment by chelleadel** [comment]
Hi! I tried it out and it works! S or T can be any objects and optional are container objects which wrap around any objects declared between <>.

**Comment by keithcharlesc** [comment]
It should work since S and T takes in anything as what the rest have mentioned above, but the implementation would probably different when writing method calls to retrieve values or for computation

**Comment by cheekang123** [comment]
It should probably work. S and T can be any type

**Comment by arnav0401** [comment]
Yes, it would work as Optionals just act as wrapper classes that wrap around any type. Think of an optional as a kind of parcel. Before you open it (or “unwrap” in the language of options) you won't know if it contains something or nothing. 

S & T are not limited to anything. You can be more specific by stating <Integer> or <String>
In the end, it all depends on how you intend the design of the code to be!

**Comment by NIL-99** [comment]
I have tried it and it works, as S and T can be used to reference any type. Furthermore, Optionals are objects that wrap around another object delcared within the arrow brackets <>, thus it is possible.

**Comment by Hanyu0118** [comment]
It will work if you initialize the Optional type like this. Also, when you call new Pair, you need to add types, like "new Pair<S,T>(Optional.ofNullable(...) , Optional.of(...))".
<img width="390" alt="截屏2021-04-18 下午5 58 59" src="https://user-images.githubusercontent.com/77627280/115141675-3f142c00-a070-11eb-9b2c-8c150ff9e187.png">

**Comment by mltnsia** [comment]
Since S and T can be any Object, and since optional is an object, it should work!

**Comment by ydymegan** [comment]
S and T can take any form so long as it is an Object. Since optional is an object, it will work

**Comment by Martini-Chwang** [comment]
S and T are objects, and so is Optional. Since Optional is an object (that wraps the S and T objects in this case), it'll work

**Comment by Joeltan15** [comment]
The constructor should be able to take in Optionals.  In which case object 1 and 2 are both optionals but containing different objects.

**Comment by garionKang62** [comment]
Yes, to directly answer your question, Optional is a type, just like Integer, Double, String, etc.
new Pair(Optional.ofNullable(...) , Optional.of(...)) is valid, although as others have also said, you should specify the typing of Pair.
S and T represent objects and they are currently not limited, so they can take in any object. You can limit them should you choose with the use of wildcards.

**Comment by utsavrakshit** [comment]
S & T is an object, hence, by virtue of that it is Optional. Thus, it will work
